dash_auth/group_protection.py

```diff
@@ -0,0 +1,198 @@
+        but does not have the right permissions.
+        It defaults to unauthenticated_output when not set.
+        Note: needs to be a function with no argument or
+        a collection of Dash components.
+    :param groups: List of authorized user groups. If no groups are passed,
+        the decorator will only check whether the user is authenticated.
+    :param groups_key: Groups key in the user data saved in the Flask session
+        e.g. session["user"] == {"email": "a.b@mail.com", "groups": ["admin"]}
+    :param groups_str_split: Used to split groups if provided as a string
+    :param check_type: Type of check to perform.
+        Either "one_of", "all_of" or "none_of"
+    """
+    if unauthenticated_output is None:
+        unauthenticated_output = ""
+
+    if missing_permissions_output is None:
+        missing_permissions_output = unauthenticated_output
+
+    def decorator(output: OutputVal):
+        def wrap(*args, **kwargs):
+            def process_output(output, *args, **kwargs):
+                if isinstance(output, Callable):
+                    return output(*args, **kwargs)
+                return output
+
+            authorized = check_groups(
+                groups=groups,
+                groups_key=groups_key,
+                groups_str_split=groups_str_split,
+                check_type=check_type,
+            )
+            if authorized is None:
+                return process_output(unauthenticated_output)
+            if authorized:
+                return process_output(output, *args, **kwargs)
+            return process_output(missing_permissions_output)
+
+        if isinstance(output, Callable):
+            return wrap
+        return wrap()
+
+    return decorator
+
+
+def protected_callback(
+    *callback_args,
+    groups: List[str] = None,
+    groups_key: str = "groups",
+    groups_str_split: str = None,
+    check_type: CheckType = "one_of",
+    **callback_kwargs,
+) -> Callable:
+    """Protected Dash callback.
+
+    :param **: all args and kwargs passed to a dash callback
+    :param groups: List of authorized user groups
+    :param groups_key: Groups key in the user data saved in the Flask session
+        e.g. session["user"] == {"email": "a.b@mail.com", "groups": ["admin"]}
+    :param groups_str_split: Used to split groups if provided as a string
+    :param check_type: Type of check to perform.
+        Either "one_of", "all_of" or "none_of"
+    """
+
+    def decorator(func):
+        def prevent_unauthenticated():
+            logging.info(
+                "A user tried to run %s without being authenticated.",
+                func.__name__,
+            )
+            raise PreventUpdate
+
+        def prevent_unauthorised():
+            logging.info(
+                "%s tried to run %s but did not have the right permissions.",
+                session["user"]["email"],
+                func.__name__,
+            )
+            raise PreventUpdate
```

Comment:
OIC `protected_callback` is using `protected` internally - my comment re: returning Dash compoents from these functions still stands, but also I imagine people using `protected_callback` may also want custom responses. If we did that, one thing that occurs to me is they might want to use one or more of the callback args in that function, which they can't get directly since this function takes no args, but they should have access to `callback_context` from in here, right?

Reply:
The way I was using this myself was really just to stop callbacks from processing further but I get your point. I added the possibility to pass the unauthenticated_output and missing_permissions_output to the protected_callback, defaulting to the prevent_x if nothing is passed.

Yes the callback context will be available in those function calls. Added it to the docstring for discoverability.